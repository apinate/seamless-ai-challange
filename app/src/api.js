import * as queryString from 'query-string';

const baseUri = process.env.REACT_APP_SERVER_URL;

const headers = new Headers();
headers.append('Accept', 'application/json');

const basicRequest = {
  headers,
  method: 'GET',
  mode: 'cors',
};

const handleErrors = (response) => {
  if (!response.ok) {
    throw new Error(response.statusText);
  }
  return response.json();
};

export const findCompaniesDomains = async (companies) => {
  const queryParams = queryString.stringify({ companies }, { arrayFormat: 'bracket' });
  return fetch(`${baseUri}/domains?${queryParams}`, basicRequest).then(handleErrors);
};
