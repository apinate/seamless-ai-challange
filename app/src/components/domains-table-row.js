import React from 'react';
import PropTypes from 'prop-types';

const DomainsTableRow = ({ data: { name, domain } }) => (
  <tr>
    <td >{name}</td>
    <td>{domain}</td>
  </tr>
);

DomainsTableRow.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    domain: PropTypes.string.isRequired,
  }).isRequired,
};

export default DomainsTableRow;
