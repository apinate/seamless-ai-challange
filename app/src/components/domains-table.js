import React from 'react';
import PropTypes from 'prop-types';

import DomainsTableRow from './domains-table-row';

const DomainsTable = (props) => {
  const renderRow = data => <DomainsTableRow key={data.name} data={data} />;

  return (
    <div className="domains-table">
      {props.companies.map(renderRow)}
    </div>
  );
};

DomainsTable.propTypes = {
  companies: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    domain: PropTypes.string.isRequired,
  }).isRequired).isRequired,
};

export default DomainsTable;
