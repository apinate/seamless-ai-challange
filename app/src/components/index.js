import DomainsTable from './domains-table';
import SearchBar from './search-bar';

export { DomainsTable, SearchBar };
