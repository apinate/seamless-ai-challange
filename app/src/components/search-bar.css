.search-icon-addon {
  background: transparent;
}

.search-input-container {
  margin: 0;
  padding: 0;
}
