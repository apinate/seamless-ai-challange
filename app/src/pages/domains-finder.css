.error-message {
  height: 40px;
}
