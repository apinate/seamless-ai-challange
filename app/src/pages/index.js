import DomainsFinder from './domains-finder';

export { DomainsFinder };
