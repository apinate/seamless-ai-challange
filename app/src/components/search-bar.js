import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './search-bar.css';

class SearchBar extends Component {
  state = { value: '' };

  onChange = ({ target: { value } }) => {
    this.setState({ value });
  }

  onClick = () => {
    const { queryExtractor, onSearch } = this.props;
    const queryData = (queryExtractor && queryExtractor(this.state.value)) || this.state.value;

    if (this.validate(queryData)) {
      onSearch(queryData);
    }
  }

  get hasError() {
    return this.state.error;
  }

  validate(queryData) {
    let error = '';
    try {
      const { validation } = this.props;
      if (validation) {
        validation(queryData);
      }
    } catch (err) {
      error = err.message;
    }
    this.setState({ error });

    return !error;
  }

  classNamesOnError(classNames) {
    return this.hasError ? classNames : '';
  }

  render() {
    return (
      <div className="mt-3">
        <div className="d-flex flex-wrap flex-row justify-content-between align-content-center">
          <div className={`form-group search-input-container col-lg-10 ${this.classNamesOnError('has-danger')}`}>
            <div className="input-group search-input-group">
              <span className="input-group-addon search-icon-addon">
                <i className="fa fa-search" aria-hidden="true" />
              </span>
              <input
                placeholder={this.props.searchPlaceholder}
                id="search-input"
                type="search"
                value={this.state.value}
                onChange={this.onChange}
                className="form-control"
              />
            </div>
          </div>

          <button className="btn btn-primary" onClick={this.onClick}>Find Domains</button>
        </div>
        <div className={`text-danger ${this.hasError ? '' : 'invisible'}`} >
          <small>{this.state.error || 'No Error'}</small>
        </div>
      </div>
    );
  }
}

SearchBar.propTypes = {
  searchPlaceholder: PropTypes.string.isRequired,
  onSearch: PropTypes.func.isRequired,
  validation: PropTypes.func,
  queryExtractor: PropTypes.func,
};

SearchBar.defaultProps = {
  validation: _ => null,
  queryExtractor: value => value,
};

export default SearchBar;
