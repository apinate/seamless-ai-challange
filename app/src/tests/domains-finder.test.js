import React from 'react';
import renderer from 'react-test-renderer';
import enzyme from 'enzyme';

import * as api from '../api';
import { DomainsFinder } from '../pages';
import { SearchBar } from '../components';


test('DomainsFinder renders correctly with no danger', () => {
  const component = renderer.create(<DomainsFinder />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('DomainsFinder shows error for no query provided', () => {
  const component = enzyme.mount(<DomainsFinder />);

  const expectedErrorMessage = 'No search query provided.';
  component.find(SearchBar).prop('onSearch')('');
  expect(component.state('queryError')).toEqual(expectedErrorMessage);
});

test('DomainsFinder shows error for exceeded company names', () => {
  const component = enzyme.shallow(<DomainsFinder />);

  const queryString = new Array(26).fill(0).map((_, i) => i).join(',');
  const expectedErrorMessage = 'Requested Company Domains Exceeds the Limit of 25';
  component.find(SearchBar).prop('onSearch')(queryString);
  expect(component.state('queryError')).toEqual(expectedErrorMessage);
});


test('DomainsFinder state should contain companies details on search', async () => {
  const component = enzyme.shallow(<DomainsFinder />);

  api.findCompaniesDomains = jest.fn().mockImplementation((names) => {
    return new Promise(resolve => resolve(names.map(name => ({ name, domain: `${name}.com` }))));
  });

  const queryString = 'adidas, nike';
  return component.find(SearchBar).prop('onSearch')(queryString).then(() => {
    expect(component.state('queryError')).toBeFalsy();
    expect(component.state('hasServerError')).toBeFalsy();

    const companies = component.state('companies');
    expect(companies).toHaveLength(2);
    expect(companies).toContainEqual({ name: 'adidas', domain: 'adidas.com' });
    expect(companies).toContainEqual({ name: 'nike', domain: 'nike.com' });
  });
});

test('DomainsFinder state should flag server error when findCompaniesDomains throws an error', async () => {
  const component = enzyme.shallow(<DomainsFinder />);

  api.findCompaniesDomains = jest.fn().mockImplementation((_) => {
    throw new Error();
  });

  const queryString = 'adidas, nike';
  return component.find(SearchBar).prop('onSearch')(queryString).then(() => {
    expect(component.state('queryError')).toBeFalsy();
    expect(component.state('hasServerError')).toBeTruthy();

    const companies = component.state('companies');
    expect(companies).toHaveLength(0);
  });
});

