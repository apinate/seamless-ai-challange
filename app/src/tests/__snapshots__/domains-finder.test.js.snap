// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainsFinder renders correctly with no danger 1`] = `
<div
  className="container"
>
  <form
    className="mt-3 d-flex flex-wrap flex-row justify-content-between align-content-center"
    onSubmit={[Function]}
  >
    <div
      className="form-group search-input-container col-lg-10 false"
    >
      <div
        className="input-group search-input-group"
      >
        <span
          className="input-group-addon search-icon-addon"
        >
          <i
            aria-hidden="true"
            className="fa fa-search"
          />
        </span>
        <input
          className="form-control"
          id="search-input"
          onChange={[Function]}
          placeholder="Enter company names separated by comma (e.g. 'company1, company2')"
          type="search"
          value=""
        />
      </div>
    </div>
    <button
      className="btn btn-primary"
      type="submit"
    >
      Find Domains
    </button>
  </form>
  <div
    className="p-1 error-message"
    role="alert"
  />
</div>
`;
